### PR TITLE
Adds steaming/updates endpoint to cafe to support real-time updates on clients

### PR DESCRIPTION
**Pull request by carsonfarmer**
To support an arbitrary number of listeners to thread updates, also adds a broadcasting system for
'channels'. Currently, only threadUpdates uses this system, but easy to add others.

### CODE REVIEW

**Comment by carsonfarmer** [comment]
Note, currently the stream returns the block information. This isn't all that useful from a client perspective. So once we have a nicer API for file objects, we should move to use that.

**Comment by carsonfarmer** [comment]
No idea what happened with that failed test @sanderpick will try to take a look later...

File: core/api.go [hunk 1]
```diff
@@ -64,6 +66,14 @@ func (a *api) Start() {
 		g.Writer.WriteHeader(http.StatusNoContent)
 	})
 
+	// Allows all origins
+	// TODO: Do not use this in production, needs to be configurable
+	config := cors.DefaultConfig()
```
Comment (carsonfarmer) [inline core/api.go]: This should probably really be part of a different PR... required to use any of our api commands in CORS context.

File: core/api.go [hunk 2]
```diff
@@ -88,8 +98,9 @@ func (a *api) Start() {
 		threads.GET("/:id", a.getThreads)
 		threads.DELETE("/:id", a.rmThreads)
 		threads.POST("/:id/files", a.addThreadFiles)
+		threads.GET("/:id/updates", a.streamThreads)
```
Comment (carsonfarmer) [inline core/api.go]: does this seem like a reasonable endpoint?

File: core/api.go [hunk 3]
```diff
@@ -88,8 +98,9 @@ func (a *api) Start() {
...
 
-		files := v0.Group("/files")
+    files := v0.Group("/files")
```
Comment (carsonfarmer) [inline core/api.go]: auto-formatting seems off here?

File: core/api.go [hunk 4]
```diff
@@ -135,12 +146,13 @@ func (a *api) Start() {
 
 // Stop stops the http api
 func (a *api) Stop() error {
-	ctx, cancel := context.WithCancel(context.Background())
+	// Use timeout to force a deadline
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```
Comment (carsonfarmer) [inline core/api.go]: useful for channels that take too long to close

File: core/api_threads.go [hunk 5]
```diff
@@ -121,3 +122,86 @@ func (a *api) rmThreads(g *gin.Context) {
...
+
+func addBlockInfo(a *api, update ThreadUpdate) (ThreadUpdate, error) {
+	block := update.Block
+	username := a.node.ContactUsername(block.AuthorId)
+
+	var info interface{}
+	switch update.Block.Type {
+	case repo.FilesBlock:
+		info, _ = a.node.threadFile(&update.Block)
+	case repo.CommentBlock:
```
Comment (carsonfarmer) [inline core/api_threads.go]: seems overly verbose? better way to do this?

File: core/broadcaster.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
+// Package core/broadcast implements multi-listener broadcast channels.
```
Comment (carsonfarmer) [inline core/broadcaster.go]: Left this in core. Should it be its own package?

File: core/main.go [hunk 7]
```diff
@@ -265,7 +265,7 @@ func (t *Textile) Start() error {
 
 	t.online = make(chan struct{})
 	t.updates = make(chan Update, 10)
-	t.threadUpdates = make(chan ThreadUpdate, 10)
+	// t.threadUpdates = make(chan ThreadUpdate, 10)
```
Comment (carsonfarmer) [inline core/main.go]: can be removed, left for review... should be gone. broadcaster is lazily instantiated, so don't need this

File: core/main.go [hunk 8]
```diff
@@ -448,8 +448,8 @@ func (t *Textile) UpdateCh() <-chan Update {
 }
 
 // ThreadUpdateCh returns the thread update channel
-func (t *Textile) ThreadUpdateCh() <-chan ThreadUpdate {
-	return t.threadUpdates
+func (t *Textile) ThreadUpdateCh() *Listener {
```
Comment (carsonfarmer) [inline core/main.go]: this now returns a listener. should the method be renamed to reflect this change?

File: core/thread.go [hunk 9]
```diff
@@ -40,9 +40,10 @@ var ErrBlockNotFile = errors.New("block is not a file")
 
 // ThreadUpdate is used to notify listeners about updates in a thread
 type ThreadUpdate struct {
-	Block      repo.Block `json:"block"`
-	ThreadId   string     `json:"thread_id"`
-	ThreadName string     `json:"thread_name"`
+	Block      repo.Block  `json:"block"`
+	ThreadId   string      `json:"thread_id"`
+	ThreadName string      `json:"thread_name"`
+	Info       interface{} `json:"info,omitempty"`
```
Comment (carsonfarmer) [inline core/thread.go]: info contains useful, human-readable, thread update information

File: core/threads_files.go [hunk 10]
```diff
@@ -43,6 +43,20 @@ type ThreadLikeInfo struct {
 	Username string    `json:"username,omitempty"`
 }
 
+type ThreadJoinInfo struct {
```
Comment (carsonfarmer) [inline core/threads_files.go]: copy-paste here... do we need these? reuse other interface perhaps?

File: mill/image_resize.go [hunk 11]
```diff
@@ -133,8 +133,15 @@ func encodeImage(reader io.Reader, format Format, width int, quality int) (*byte
 		}
 
 		resized := imaging.Resize(img, width, 0, imaging.Lanczos)
-		if err = jpeg.Encode(buff, resized, &jpeg.Options{Quality: quality}); err != nil {
-			return nil, nil, err
+
+		if format == PNG {
```
Comment (carsonfarmer) [inline mill/image_resize.go]: quick fix to make sure we encode pngs correctly

File: mobile/mobile.go [hunk 12]
```diff
@@ -177,10 +177,11 @@ func (m *Mobile) Start() error {
 		}()
 
 		// subscribe to thread updates
+		listener := m.node.ThreadUpdateCh()
```
Comment (carsonfarmer) [inline mobile/mobile.go]: new thread update api usage here as well

File: textile.go [hunk 13]
```diff
@@ -394,24 +394,30 @@ func startNode() error {
 		}
 	}()
 
-	// subscribe to thread updates
+	// Subscribe to thread updates
+	listener := node.ThreadUpdateCh()
```
Comment (carsonfarmer) [inline textile.go]: also here

**Comment by carsonfarmer** [comment]
Yeh hmmm, re-ran them all locally, looks good here?

File: cmd/threads.go [hunk 14]
```diff
@@ -300,6 +304,72 @@ func callRmThreads(args []string) error {
...
+
+func (x *streamThreadsCmd) Short() string {
+	return "Steam thread updates"
+}
+
+func (x *streamThreadsCmd) Long() string {
+	return `
+Streams all thread update types.
+Use the --events option to emit Server-Sent Events (SSEvent), otherwise, emit JSON responses. 
+SSEvents enable browsers/clients to consume the stream using EventSource.
```
Comment (sanderpick) [inline cmd/threads.go]: hell yeah

File: core/api.go (hunk 1 above)
Comment (sanderpick) [reply core/api.go]: cool, can you open an issue to make this part of the textile config file?

File: core/api.go (hunk 2 above)
Comment (sanderpick) [reply core/api.go]: lgtm

File: core/api.go (hunk 4 above)
Comment (sanderpick) [reply core/api.go]: perf

File: core/api.go (hunk 3 above)
Comment (sanderpick) [reply core/api.go]: huh, yeah

File: core/contacts.go [hunk 15]
```diff
@@ -75,7 +75,7 @@ func (t *Textile) ContactUsername(id string) string {
 	if id == "" {
 		return ""
 	}
-	username = id[len(id)-7:]
+	username = id[len(id)-8:]
```
Comment (sanderpick) [inline core/contacts.go]: it's 7 everywhere else... let's keep that

File: core/main.go (hunk 7 above)
Comment (sanderpick) [reply core/main.go]: nice! kill it

File: core/main.go (hunk 8 above)
Comment (sanderpick) [reply core/main.go]: sounds right... `GetThreadUpdateListener`?

File: core/broadcaster.go (hunk 6 above)
Comment (sanderpick) [reply core/broadcaster.go]: cool stuff. yeah, seems like it could be it's own package

File: mill/image_resize.go (hunk 11 above)
Comment (sanderpick) [reply mill/image_resize.go]: sweet

File: core/threads_files.go (hunk 10 above)
Comment (sanderpick) [reply core/threads_files.go]: yeah, hmmm. I see what you mean. I just added username to the BlockInfo object, which is available via Textile.BlockInfo(blockId)... how about just replaceing `Block` on ThreadUpdate w/ `BlockInfo`, and then only in the case where you want to inflate the files update info tack on the `Info` field?

File: core/api_threads.go (hunk 5 above)
Comment (sanderpick) [reply core/api_threads.go]: hhmmm, yeah. see comment below about `BlockInfo`

**Comment by sanderpick** [comment]
failure is this annoying map concurrent read thing, happens once in awhile (only in tests w/ multiple nodes trying to access the ipfs go-logging module)

File: core/api.go (hunk 1 above)
Comment (carsonfarmer) [reply core/api.go]: added

File: core/api.go (hunk 3 above)
Comment (carsonfarmer) [reply core/api.go]: fixed in upcoming push

File: core/contacts.go (hunk 15 above)
Comment (carsonfarmer) [reply core/contacts.go]: oh hah, just realized what i did there... nv, reverting 😬

**Comment by carsonfarmer** [comment]
aside from that annoying test issue, this should be GTG now sander

File: core/thread.go [hunk 16]
```diff
@@ -441,7 +442,18 @@ func (t *Thread) indexBlock(commit *commitResult, blockType repo.BlockType, targ
...
-	t.pushUpdate(*index)
+	info := &BlockInfo{
+		Id:       index.Id,
+		ThreadId: index.ThreadId,
+		AuthorId: index.AuthorId,
+		Type:     index.Type.Description(),
+		Date:     index.Date,
+		Parents:  index.Parents,
+		Target:   index.Target,
+		Body:     index.Body,
```
Comment (sanderpick) [inline core/thread.go]: if you do `t.BlockInfo(index.Id)` you'll get the info w/ `Username` hydrated

File: core/api_threads.go [hunk 17]
```diff
@@ -121,3 +122,65 @@ func (a *api) rmThreads(g *gin.Context) {
...
+func addBlockInfo(a *api, update ThreadUpdate) (ThreadUpdate, error) {
+	block := update.Block
+	username := a.node.ContactUsername(block.AuthorId)
+
+	var info interface{}
+	switch update.Block.Type {
+	case "FILES":
+		info, _ = a.node.ThreadFile(update.Block.Id)
+	case "COMMENT":
+		info = ThreadCommentInfo{
```
Comment (sanderpick) [inline core/api_threads.go]: w/ username on block info, you could kill this too.. redundant

File: core/api_threads.go [hunk 18]
```diff
@@ -121,3 +122,65 @@ func (a *api) rmThreads(g *gin.Context) {
...
+}
+
+func addBlockInfo(a *api, update ThreadUpdate) (ThreadUpdate, error) {
+	block := update.Block
+	username := a.node.ContactUsername(block.AuthorId)
+
+	var info interface{}
+	switch update.Block.Type {
+	case "FILES":
+		info, _ = a.node.ThreadFile(update.Block.Id)
```
Comment (sanderpick) [inline core/api_threads.go]: you could `log.Errorf("error getting thread file: %s", err)` here

File: core/thread.go (hunk 16 above)
Comment (carsonfarmer) [reply core/thread.go]: didn't go this route, but as per our slack convo, merged master and added username to block info

**Comment by carsonfarmer** [comment]
pushed up a sneaky fix, will merge when done checks
